server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -181,18 +169,87 @@ remoteSchemaObject schemaDoc defn@(G.ObjectTypeDefinition description name inter
       = True -- TODO write appropriate check (may require saving 'possibleTypes' in Syntax.hs)
     validateSubTypeDefinition _ _ = False
 
+-- | helper function to get a parser of an object with it's name
+--   This function is called from 'remoteSchemaInterface' and
+--   'remoteSchemaObject' functions. Both of these have a slightly
+--   different implementation of 'getObject', which is the
+--   reason 'getObject' is an argument to this function
+getObjectParser
+  :: forall n m
+  . (MonadSchema n m, MonadError QErr m)
+  => SchemaIntrospection
+  -> (G.Name -> m G.ObjectTypeDefinition)
+  -> G.Name
+  -> m (Parser 'Output n (Name, (SelectionSet NoFragments G.Name)))
+getObjectParser schemaDoc getObject objName = do
+  obj <- remoteSchemaObject schemaDoc =<< getObject objName
+  return $ (objName,) <$> obj
+
+{- Note [Querying remote schema interfaces]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+When querying Remote schema interfaces, we need to re-construct
+the incoming query to be compliant with the upstream remote.
+We need to do this because the `SelectionSet`(s) that are
+inputted to this function have the fragments (if any) flattened.
+(Check `flattenSelectionSet` in 'Hasura.GraphQL.Parser.Collect' module)
+The `constructInterfaceSelectionSet` function makes a valid interface query by:
+1. Getting the common interface fields in all the selection sets
+2. Remove the common fields obtained in #1 from the selection sets
+3. Construct a selection field for every common interface field
+4. Construct inline fragments for non-common interface fields
+   using the result of #2 for every object
+5. Construct the final selection set by combining #3 and #4
+
+Example: Suppose an interface 'Character' is defined in the upstream
+and two objects 'Human' and 'Droid' implement the 'Character' Interface.
+
+Suppose, a field 'hero' returns 'Character'.
+
+{
+   hero {
+     id
+     name
+     ... on Droid {
+       primaryFunction
+     }
+     ... on Human {
+       homePlanet
+     }
+   }
+}
+
+After parsing, the above query, the selection set is modified to:
+
+{
+   hero {
+     id
+     name
+     primaryFunction
+     homePlanet
+   }
+}
```

Comment:
This isn't quite right, and has to be phrased more carefully, I think.

1. I don't think any code in the current PR would ever see a selection set with 4 fields as in the example above.
2. Selection sets aren't mutable, so we can't say it's "modified". You can say that some value is returned by some parser, but then it'd be valuable to specify which one.

The point is that when parsing the interface selection set for the `hero` field, we parse the selection set twice: once for the `Droid` object type, which would be passed a selection set containing the fields `id`, `name`, and `primaryFunction`, and once for the `Human` object type, which would be passed a selection set containing the fields `id`, `name` and `homePlanet`. The result of the interface selection set parsing would then be the results of the parsing of the object types when passed their corresponding _restricted_ (and flattened) selection sets. So `Droid` might wish to report a selection set of `{id name primaryFunction}`, and `Human` might report a selection set of `{id name homePlanet}`.

Reply:
Thank you for the explanation, I was a bit confused because I did a `Debug.trace` on the output of `P.selectionSetInterface` and thought it returns a selection set with the 4 fields as I mentioned earlier. I've modified the note based on your explanation.